File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # hashset = set()
        # while head:
        #     if head in hashset:
        #         return True
        #     hashset.add(head)
        #     head = head.next
        # return False

        fast, slow = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True
        return False

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode RemoveNthFromEnd(ListNode head, int n) {
        ListNode sz = head;
        int len = 0;
        while (sz != null) {
            len += 1;
            sz = sz.next;
        }
        if (len == n) {
            return head.next;
        }

        ListNode rm = head;
        for (int i = 1; i < len - n; i++) {
            rm = rm.next;
        }
        ListNode temp = rm.next;
        rm.next = temp.next;
        return head;
    }
}
